control-plane-operator/controllers/hostedcontrolplane/oauth/idp_convert.go

```diff
@@ -0,0 +1,570 @@
+					File: idpVolumeMounts.SecretPath(i, openIDConfig.ClientSecret.Name, "client-secret", configv1.ClientSecretKey),
+				},
+			},
+			ExtraScopes:              openIDConfig.ExtraScopes,
+			ExtraAuthorizeParameters: openIDConfig.ExtraAuthorizeParameters,
+			URLs:                     *urls,
+			Claims: osinv1.OpenIDClaims{
+				// There is no longer a user-facing setting for ID as it is considered unsafe
+				ID:                []string{configv1.UserIDClaim},
+				PreferredUsername: openIDConfig.Claims.PreferredUsername,
+				Name:              openIDConfig.Claims.Name,
+				Email:             openIDConfig.Claims.Email,
+			},
+		}
+
+		// openshift CR validating in kube-apiserver does not allow
+		// challenge-redirecting IdPs to be configured with OIDC so it is safe
+		// to allow challenge-issuing flow if it's available on the OIDC side
+		challengeFlowsAllowed, err := checkOIDCPasswordGrantFlow(
+			kclient,
+			urls.Token,
+			openIDConfig.ClientID,
+			namespace,
+			openIDConfig.CA,
+			openIDConfig.ClientSecret,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error attempting password grant flow: %v", err)
+		}
+		data.challenge = challengeFlowsAllowed
+
+	case configv1.IdentityProviderTypeRequestHeader:
+		requestHeaderConfig := providerConfig.RequestHeader
+		if requestHeaderConfig == nil {
+			return nil, fmt.Errorf(missingProviderFmt, providerConfig.Type)
+		}
+
+		data.provider = &osinv1.RequestHeaderIdentityProvider{
+			LoginURL:                 requestHeaderConfig.LoginURL,
+			ChallengeURL:             requestHeaderConfig.ChallengeURL,
+			ClientCA:                 idpVolumeMounts.ConfigMapPath(i, requestHeaderConfig.ClientCA.Name, "ca", corev1.ServiceAccountRootCAKey),
+			ClientCommonNames:        requestHeaderConfig.ClientCommonNames,
+			Headers:                  requestHeaderConfig.Headers,
+			PreferredUsernameHeaders: requestHeaderConfig.PreferredUsernameHeaders,
+			NameHeaders:              requestHeaderConfig.NameHeaders,
+			EmailHeaders:             requestHeaderConfig.EmailHeaders,
+		}
+		data.challenge = len(requestHeaderConfig.ChallengeURL) > 0
+		data.login = len(requestHeaderConfig.LoginURL) > 0
+
+	default:
+		return nil, fmt.Errorf("the identity provider type '%s' is not supported", providerConfig.Type)
+	} // switch
+
+	return data, nil
+}
+
+// discoverOpenIDURLs retrieves basic information about an OIDC server with hostname
+// given by the `issuer` argument
+func discoverOpenIDURLs(kclient crclient.Client, issuer, key, namespace string, ca configv1.ConfigMapNameReference) (*osinv1.OpenIDURLs, error) {
+	issuer = strings.TrimRight(issuer, "/") // TODO make impossible via validation and remove
+	wellKnown := issuer + "/.well-known/openid-configuration"
+
+	cacheValue, inCache := openIDURLsCache.Get(wellKnown)
+	if inCache {
+		return cacheValue.(*osinv1.OpenIDURLs), nil
+	}
+
+	reqCtx, cancel := context.WithTimeout(context.Background(), externalHTTPRequestTimeout)
```

Comment:
I think you should pass the context down all the way from the reconciler to this point so we have consistent context awareness for tracing, timeouts, etc.